semgrep-core/tests/OTHER/rules/metavar_regex.yaml

```diff
@@ -5,8 +5,8 @@ rules:
   match:
     and:
```

Comment:
@kopecs in imperative languages I see these names for equivalent operations:

| Language                | AND operation                    | OR operation                     | Notes                                                                                                 |
|-------------------------|----------------------------------|----------------------------------|-------------------------------------------------------------------------------------------------------|
| Python                  | `all(predicate for i in [a, b])` | `any(predicate for i in [a, b])` |                                                                                                       |
| JavaScript / TypeScript | `[a, b].every(predicate)`        | `[a, b].some(predicate)`         |                                                                                                       |
| Java                    | `[a, b].allMatch(predicate)`     | `[a, b].anyMatch(predicate)`     | (this is with something called Stream? I dunno java)                                                  |
| Ruby                    | `[a, b].all? {predicate}`        | `[a, b].any? {predicate}`        |                                                                                                       |
| PHP                     |                                  |                                  | has nothing lol                                                                                       |
| Golang                  | `SomeBy({a, b}, predicate)`      | `EveryBy({a, b}, predicate)`     | also has nothing, but [a popular utility belt library](https://github.com/samber/lo) uses these names |
| Kotlin                  | `arrayOf(a, b).any(predicate)`   | `arrayOf(a, b).all(predicate)`   |                                                                                                       |
| Scala                   | `Vector(a, b).exists(predicate)` | `Vector(a, b).forall(predicate)` |                                                                                                       |

In all of these languages, the `and` and `or` keywords are used only as binary operators.

Since these are the top eight programming languages our users scan (multiple hundreds of times more popular than all other languages combined), would you be open to renaming `and:/or:` in our rule schema to `all:/any:`?

Reply:
I like `all` and `any`. They appear in Haskell too. They also read more nicely with `match`. 
```
match:
  any:
```
reads like "match any of these patterns"

Reply:
+1 to all/any

Reply:
I also found the original discussion when any/all were being added to Python where this naming vs or/and was discussed:

https://mail.python.org/archives/list/python-dev@python.org/thread/K33F5RHPA6UTPU3SIYUMTVM4B7P3JUGC/#N5EBG3QLC4WS3H4Z4WLRFEBXHI4QLQYA

https://www.artima.com/forums/flat.jsp?forum=106&thread=98196

Reply:
Yeah, we talked a little bit about names, and I'm not in love with the exact wording we've got in here.

> in imperative languages I see these names for equivalent operations on arrays:

I don't think any of the examples you have here are analogous---they're all about applying one fixed predicate to array elements, whereas this is essentially a combinator on an array of predicates.

> In all of these languages, the and and or keywords are used only as binary operators.

In any event, I don't think anyone is particularly confused if these are binary operations or not. I've not seen any confusion in our current syntax relating to this, and `pattern-either` is far more indicative of only being a binary operator imo; more broadly, for associative operations I don't think this is a concern---I've never heard the response to "add these 3 things together" be "wait a second! addition is a binary operation, so you can't add 3 things".

> would you be open to renaming and/or in our rule schema to all/any?

and/or vs all/any is mostly the same to me, with maybe a slight preference for the former because I can't read good and they're a bit more visually distinct. If others have a strong preference, happy to change it.

The issue I have with both of these options is that they're not super reflective of the actual operation performed: `and`/`patterns`/whatever doesn't actually (despite what our documentation claims) perform a boolean and operation, nor does it check that "all" of the patterns match exactly---it results in the smallest range in the target file where all the child patterns match some range which is contained/contained by all of the other combined ranges. This is less of an issue for `or`/`patterns-either`/whatever, because its just choice.

Reply:
>and/or vs all/any is mostly the same to me, maybe a slight preference for the former because I can't read good and they're a bit more visually distinct. If others have a strong preference, happy to change it.

every/some are a visually distinct option, but those words sound more obscure and arbitrary to me, plus it's extra characters

I also wanna mention ElasticSearch's query DSL, which is the closest reference I can come up with, since our rule syntax is sort of a query DSL as well. They use what I think is a frankly absurd set of keywords: [filter/must/should/must_not](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html). I wouldn't really wanna copy this, it was always annoying me as a user of ES. To be fair, they have an extra requirement of also using this for scoring results, which might have to do with the weird keywords, but still.

For the record I never issues with misreading any/all in python code; they're not all that similar in fixed width font. It is worthwhile to think about accessibility for dyslexic users though. 💡 I wonder if it'd be a worthwhile addition to our LSP server to add `&&` and `||` as inlay hints between the array elements?

Reply:
Meta comment: I dislike the idea that we're calling this bikeshedding. I think this is a serious language design question that's worth discussing if we're gonna release a Semgrep v1.

Reply:
And if a PR trying to make our pattern operator names better is not the right place for discussing better pattern operator names, I'm not sure what is :D

Reply:
worldssmallestviolin.avi

Reply:
I agree that this is important; I don't think we should decide by talking and theorizing. In terms of actual feedback, I was there when Spencer was interviewed and he understood and/or without any trouble. Obviously we need some more data points for both options.

Reply:
My point though is not that all/any is more understandable than and/or, but that it's more readable (reduces cognitive load while reading and writing rules). So we'd need to design the experiment around that, and the anecdote about spencer is not too relevant.

Reply:
Can you disambiguate what you see as the difference between "understandable" and "readable" here @underyx? I don't understand how you're differentiating them with respect to cognitive load.

Some other options to throw in the mix, just so we don't collapse to just these two ideas too early:
- For `patterns`:
  - `common` (because we're really getting the range common to all the patterns)
  - `intersect` (a bit literal, but we're \~kinda~ doing an intersection, with the constraint that ranges are contained within another)
- For `pattern-either`:
  - `choice` or `choose` (common in parser combinator libraries, e.g., [Parsec](https://hackage.haskell.org/package/parsec-3.1.15.1/docs/Text-Parsec-Combinator.html#v:choice); [racket's Parsack](https://docs.racket-lang.org/parsack/index.html#%28def._%28%28lib._parsack%2Fmain..rkt%29._choice%29%29); [chumsky](https://docs.rs/chumsky/latest/chumsky/primitive/fn.choice.html)), 
  - `union` (the produced set of ranges is the union of the results of the children patterns)

Reply:
sure, what I meant is:

- I expect `and/or` and `all/any` are equally understandable: 99 of 100 readers would correctly explain what a rule does (and this is what we tested on spencer)
- I expect `all/any` is more readable than `and/or`: the average time taken to correctly explain what a rule does would be shorter in the `all/any` case compared to `and/or`. I also expect the gap between the two options expands with every level of nesting that the rule has.

Reply:
To illustrate with some fake made-up data, I think if we had read access to profiling data from a test subject's brain, if we showed someone a 20-line rule which has four levels of nesting of `and`, `or`, `or`, `and` operators nested within each other, they'd comprehend the rule in 3600ms, while if we showed the same rule with `all`, `any`, `any`, `all`, they'd comprehend it in 3500ms instead.

I also think that a language that's easier to read will often also be easier to write.

Reply:
+1 to `all` and `any`.

Reply:
Isn't the main reason for Python using `all` and `any` that `and` and `or` are reserved keywords?